### PR TITLE
[Debugger] Fix crash when there is a generic struct with a field that is an enumerator.

### LINKED ISSUE #10735: mono64[80780:1186894] error: * Assertion: should not be reached at debugger-agent.c:6619
## Steps to Reproduce

1. Open https://github.com/mono/monodevelop/ solution in VS for Mac (7.8 Preview)
2. Set breakpoint to https://github.com/mono/monodevelop/blob/305bddc1e9dcd90caf25f8bed1a042efc317528b/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ProjectFileCollection.cs#L146
3. Set configuration to DebugMac and Run
4. Open a project and add some file to it
5. Rename the file
6. On breakpoint hit try to inspect `files` field, (expand Enumerator)

[monocrash.log](https://github.com/mono/mono/files/2405309/monocrash.log)
[monocrash2.log](https://github.com/mono/mono/files/2405357/monocrash2.log)

### Current Behavior

Debugger crashes (The Debugger is Busy)

### Expected Behavior

It shouldn't

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

```
Mono JIT compiler version 5.16.0.112 (2018-06/b59ae60011a Wed Aug 15 09:27:21 EDT 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(3.6.0svn-mono-master/0b3cb8ac12c)
	GC:            sgen (concurrent by default)
```

## Stacktrace

```
2018-09-21 15:36:47.153 mono64[80895:1191537] error: * Assertion: should not be reached at debugger-agent.c:6619
Stacktrace:

  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&) [0x00016] in <0242286efc4a47b69b5a84e828f56ef8>:0
  at System.Reflection.MonoMethod.Invoke (object,System.Reflection.BindingFlags,System.Reflection.Binder,object[],System.Globalization.CultureInfo) [0x0003b] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/corlib/System.Reflection/MonoMethod.cs:305
  at System.Reflection.MethodBase.Invoke (object,object[]) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/referencesource/mscorlib/system/reflection/methodbase.cs:229
  at System.Delegate.DynamicInvokeImpl (object[]) [0x000e1] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/Delegate.cs:467
  at System.MulticastDelegate.DynamicInvokeImpl (object[]) [0x00008] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/MulticastDelegate.cs:67
  at System.Delegate.DynamicInvoke (object[]) [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/Delegate.cs:412
  at GLib.Signal.ClosureInvokedCB (object,GLib.ClosureInvokedArgs) [0x0007f] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/gtk-sharp-None/glib/Signal.cs:207
  at GLib.Signal/SignalClosure.Invoke (GLib.ClosureInvokedArgs) [0x00019] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/gtk-sharp-None/glib/SignalClosure.cs:114
  at GLib.Signal/SignalClosure.MarshalCallback (intptr,GLib.Value*,uint,GLib.Value*,intptr,intptr) [0x000b9] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/gtk-sharp-None/glib/SignalClosure.cs:153
  at (wrapper native-to-managed) GLib.Signal/SignalClosure.MarshalCallback (intptr,GLib.Value*,uint,GLib.Value*,intptr,intptr) [0x00027] in <bbc097c736e649f0bbe49b81946fd886>:0
  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) Gtk.Widget.gtksharp_widget_base_key_press_event (intptr,intptr) [0x0000a] in <40f46e55eb67454ab904dc9e5644a3a1>:0
  at Gtk.Widget.OnKeyPressEvent (Gdk.EventKey) [0x00022] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/gtk-sharp-None/gtk/generated/Widget.cs:1412
  at MonoDevelop.Ide.Gui.DefaultWorkbench.OnKeyPressEvent (Gdk.EventKey) [0x00001] in /Users/user/Xamarin/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/DefaultWorkbench.cs:1244
  at Gtk.Widget.keypressevent_cb (intptr,intptr) [0x00016] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/gtk-sharp-None/gtk/generated/Widget.cs:1390
  at (wrapper native-to-managed) Gtk.Widget.keypressevent_cb (intptr,intptr) [0x00022] in <40f46e55eb67454ab904dc9e5644a3a1>:0
  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) Gtk.Application.gtk_main () [0x00008] in <40f46e55eb67454ab904dc9e5644a3a1>:0
  at Gtk.Application.Run () [0x00001] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-06/external/bockbuild/builds/gtk-sharp-None/gtk/Application.cs:145
  at MonoDevelop.Ide.IdeApp.Run () [0x00007] in /Users/user/Xamarin/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/Ide.cs:464
  at MonoDevelop.Ide.IdeStartup.Run (MonoDevelop.Ide.MonoDevelopOptions) [0x008c9] in /Users/user/Xamarin/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/IdeStartup.cs:332
  at MonoDevelop.Ide.IdeStartup.Main (string[],MonoDevelop.Ide.Extensions.IdeCustomizer) [0x000c4] in /Users/user/Xamarin/monodevelop/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide/IdeStartup.cs:740
  at MonoDevelop.Startup.MonoDevelopMain.Main (string[]) [0x00001] in /Users/user/Xamarin/monodevelop/main/src/core/MonoDevelop.Startup/MonoDevelop.Startup/MonoDevelopMain.cs:39
  at (wrapper runtime-invoke) <Module>.runtime_invoke_int_object (object,intptr,intptr,intptr) [0x00057] in <cf3b0a4af048464198115cc5e3f9ac0f>:0
```

### PR DESCRIPTION
**Pull request by thaystg**
Debugger crashes when there is a generic struct with a field that is an enumerator
Example: files.get_Current().Key
A unit test that reproduces this crash was added too.
Fixes #10735

### CODE REVIEW

**Comment by marek-safar** [comment]
Please backport down to 2018-08

File: mcs/class/Mono.Debugger.Soft/Test/dtest.cs [hunk 1]
```diff
@@ -4584,6 +4584,32 @@ Event step_out_await (string method, Event e)
 		assert_location(e, "ss_nested_arg");
 	}
 
+	[Test]
+	public void InspectFieldContent() {
+		//files.myBucket.GetEnumerator().get_Current().Key watching this generates an exception in Debugger
+		Event e = run_until("inspect_field_content");
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: Could this test be simplified ?

Comment (vargaz) [reply mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: I assume only the enum field is needed, the Enumerator etc. stuff is not.

Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: No...This works completely fine. I will try another thing...
public class TestVargaz
    {
        private Position _currentPosition;
        private enum Position
        {
            BeforeFirst
        }
        internal TestVargaz()
        {
            _currentPosition = Position.BeforeFirst;
        }
    }

Comment (vargaz) [reply mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: It needs to be a struct i think.

**Comment by thaystg** [comment]
@monojenkins build failed

File: mono/mini/debugger-agent.c [hunk 2]
```diff
@@ -5595,7 +5595,13 @@ decode_value_internal (MonoType *t, int type, MonoDomain *domain, guint8 *addr,
...
 				return ERR_INVALID_ARGUMENT;
 			}
-		} else {
+		} else if ((t->type == MONO_TYPE_GENERICINST) && 
+					mono_metadata_generic_class_is_valuetype (t->data.generic_class) &&
+					m_class_is_enumtype (t->data.generic_class->container_class)){
+			err = decode_vtype (t, domain, addr, buf, &buf, limit);
+			if (err != ERR_NONE)
+				return err;
+		} else  {
```
Comment (luhenry) [inline mono/mini/debugger-agent.c]: Nit: there is an extra space, that triggers an extra diff line.

Comment (thaystg) [reply mono/mini/debugger-agent.c]: Thanks! Removed!

**Comment by vargaz** [comment]
@monojenkins build

**Comment by lewurm** [comment]
@monojenkins backport 2018-08

**Comment by lewurm** [comment]
@monojenkins backport 2018-10

**Comment by lewurm** [comment]
@monojenkins backport 2018-12

**Pull request merged**